### PR TITLE
[2296] Add "export to SVG" action in the panel of react-flow prototype

### LINKED ISSUE #2296: Add Export as image (SVG / PNG) main toolbar action on reactflow diagrams
* [ ] **I have checked that this feature has not yet been suggested by someone else.**

### Expected behavior

Add Export as image (SVG / PNG) main toolbar action on reactflow diagrams

### Cuttings backs

...

### Potential side effects

...

### Describe alternatives you've considered

...

### PR DESCRIPTION
**Pull request by AxelRICHARD**
Bug: https://github.com/eclipse-sirius/sirius-web/issues/2296

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [x] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [x] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [x] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [x] Have the relevant issues been added to the pull request?
- [x] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [x] Have the relevant issues been added to the same project and milestone as the pull request?
- [x] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramPanel.tsx [hunk 1]
```diff
@@ -50,18 +59,32 @@ export const DiagramPanel = ({
...
   const onUnfadeAll = () => fadeDiagramElements([...getAllElementsIds()], false);
   const onUnhideAll = () => hideDiagramElements([...getAllElementsIds()], false);
 
+  const handleExport = () => {
+    const imageWidth: number = window.screen.width;
+    const imageHeight: number = window.screen.height;
+    const nodesBounds: Rect = getRectOfNodes(reactFlow.getNodes());
+    const transform: Transform = getTransformForBounds(nodesBounds, imageWidth, imageHeight, 0.5, 2);
+
+    toSvg(document.querySelector('.react-flow__viewport') as HTMLElement, {
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramPanel.tsx]: Remove _as HTMLLElement_, `document.querySelector()` already returns an `HTMLElement`

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramPanel.tsx [hunk 2]
```diff
@@ -89,6 +112,9 @@ export const DiagramPanel = ({
           <IconButton size="small" onClick={handleShare}>
             <ShareIcon />
           </IconButton>
+          <IconButton size="small" onClick={handleExport}>
+            <ImageIcon />
+          </IconButton>
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramPanel.tsx]: Can you add a tooltip like the reveal hidden and reveal faded actions?

**Review by gcoutable — CHANGES_REQUESTED** [review]

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramPanel.tsx (hunk 1 above)
Comment (AxelRICHARD) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramPanel.tsx]: On my side, `document.querySelector()` returns an `Element`

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramPanel.tsx (hunk 2 above)
Comment (AxelRICHARD) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramPanel.tsx]: Yes I can

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramPanel.tsx (hunk 1 above)
Comment (gcoutable) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramPanel.tsx]: My bad,
You can write it like this `document.querySelector<HTMLElement>('.react-flow__viewport')` then.

Comment (AxelRICHARD) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramPanel.tsx]: Ok

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramPanel.tsx [hunk 3]
```diff
@@ -50,18 +59,32 @@ export const DiagramPanel = ({
...
+    toSvg(document.querySelector<HTMLElement>('.react-flow__viewport')!, {
+      backgroundColor: '#ffffff',
+      width: imageWidth,
+      height: imageHeight,
+      style: {
+        width: imageWidth.toString(),
+        height: imageHeight.toString(),
+        transform: `translate(${transform[0]}px, ${transform[1]}px) scale(${transform[2]})`,
+      },
+    }).then(downloadImage);
```
Comment (gcoutable) [inline packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramPanel.tsx]: the '!' should not be used like this.
It is the same as doing `optional.get()` without doing `if(optional.isPresent())` before in Java.
 
Prefer something like
```typescript
const viewport: HTMLElement | null = document.querySelector<HTMLElement>('.react-flow__viewport');
if (viewport) {
  toSvg(...)
}
```

**Review by gcoutable — CHANGES_REQUESTED** [review]

File: packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramPanel.tsx (hunk 3 above)
Comment (AxelRICHARD) [reply packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/DiagramPanel.tsx]: Done

**Review by gcoutable — CHANGES_REQUESTED** [review]
I tried it on Papaya and it exported only a small part of the screen, not the whole diagram

**Comment by AxelRICHARD** [comment]
> I tried it on Papaya and it exported only a small part of the screen, not the whole diagram

It should be fixed now, thank you.

**Review by gcoutable — APPROVED** [review]

**Pull request merged**